File: web/src/lib/progress.ts
export interface VocabSkill {
  band: number
  total_words: number
  mastered_count: number
}

export interface WritingSkill {
  band: number
  sample_size: number
}

export interface ListeningSkill {
  band: number
  sample_size: number
  accuracy_by_type: Record<string, number>
}

export interface SkillBreakdown {
  vocabulary: VocabSkill
  writing: WritingSkill
  listening: ListeningSkill
}

export interface ProgressSnapshot {
  overall_band: number
  skills: SkillBreakdown
  target_band: number
  date: string | null
  generated_at: string | null
}

export interface TrendPoint {
  date: string
  overall_band: number
  vocabulary_band: number
  writing_band: number
  listening_band: number
}

export interface ProgressPrediction {
  days_ahead: number
  projected_band: number
}

export interface ProgressResponse {
  snapshot: ProgressSnapshot
  trend: TrendPoint[]
  predictions: ProgressPrediction[]
}

export function deltaFrom(
  trend: TrendPoint[],
  key: keyof TrendPoint,
): number {
  if (trend.length < 2) return 0
  const firstNonZero = trend.find(
    (p) => typeof p[key] === 'number' && (p[key] as number) > 0,
  )
  const last = trend[trend.length - 1]
  if (!firstNonZero || !last) return 0
  const start = Number(firstNonZero[key]) || 0
  const end = Number(last[key]) || 0
  return Math.round((end - start) * 10) / 10
}

export function timeToTarget(
  current: number,
  target: number,
  trend: TrendPoint[],
): string | null {
  if (current >= target) return 'Đã đạt mục tiêu'
  if (trend.length < 3) return null
  const first = trend.find((p) => p.overall_band > 0)
  const last = trend[trend.length - 1]
  if (!first || !last) return null
  const daysSpan = trend.indexOf(last) - trend.indexOf(first)
  if (daysSpan <= 0) return null
  const rate = (last.overall_band - first.overall_band) / daysSpan
  if (rate <= 0) return 'Cần tăng nhịp luyện tập'
  const remaining = (target - current) / rate
  return `Dự kiến ${Math.round(remaining)} ngày nữa`
}

File: api/models/progress.py
from datetime import datetime

from pydantic import BaseModel


class VocabSkill(BaseModel):
    band: float
    total_words: int = 0
    mastered_count: int = 0


class WritingSkill(BaseModel):
    band: float
    sample_size: int = 0


class ListeningSkill(BaseModel):
    band: float
    sample_size: int = 0
    accuracy_by_type: dict[str, float] = {}


class ReadingSkill(BaseModel):
    band: float
    sample_size: int = 0


class SkillBreakdown(BaseModel):
    vocabulary: VocabSkill
    writing: WritingSkill
    listening: ListeningSkill
    reading: ReadingSkill


class ProgressSnapshot(BaseModel):
    overall_band: float
    skills: SkillBreakdown
    target_band: float = 7.0
    date: str | None = None
    generated_at: datetime | None = None


class TrendPoint(BaseModel):
    date: str
    overall_band: float
    vocabulary_band: float
    writing_band: float
    listening_band: float
    reading_band: float = 0.0


class ProgressPrediction(BaseModel):
    days_ahead: int
    projected_band: float


class ProgressResponse(BaseModel):
    snapshot: ProgressSnapshot
    trend: list[TrendPoint]
    predictions: list[ProgressPrediction]


class ProgressHistoryResponse(BaseModel):
    items: list[TrendPoint]


class CoachingTip(BaseModel):
    id: str
    skill: str
    tip_en: str
    tip_vi: str
    action_label: str
    action_route: str


class ProgressRecommendationsResponse(BaseModel):
    week_key: str
    tips: list[CoachingTip]
    generated_at: datetime | None = None
